### ISSUE TITLE
CountDown

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#458]
I propose to add a sequence extension that provides a countdown for some last count elements of a sequence. The countdown is `null`for all but the last count of elements. The countdown value for each of the last count of elements represents the count of remaining elements, or seen another way, offset from the end of the sequence.

## Prototype

```c#
public static IEnumerable<TResult> CountDown<T, TResult>(this IEnumerable<T> source,
    int count, Func<T, int?, TResult> resultSelector)
{
    var queue = new Queue<T>();
    foreach (var item in source)
    {
        queue.Enqueue(item);
        if (queue.Count > count)
            yield return resultSelector(queue.Dequeue(), null);
    }
    while (queue.Count > 0)
        yield return resultSelector(queue.Dequeue(), queue.Count);
}
```

## Examples

```c#
Enumerable.Range(1, 5)
          .CountDown(count, (x, cd) => new { Element = x, CountDown = cd })
          .ForEach(Console.WriteLine);
```

The tables below show the output that can be expected for various values of `count` in the example above.

### When `count` = 0

| Element | CountDown |
|---------|-----------|
| 1       | null      |
| 2       | null      |
| 3       | null      |
| 4       | null      |
| 5       | null      |

### When `count` = 3

| Element | CountDown |
|---------|-----------|
| 1       | null      |
| 2       | null      |
| 3       | 2         |
| 4       | 1         |
| 5       | 0         |

### When `count` = 5

| Element | CountDown |
|---------|-----------|
| 1       | 4         |
| 2       | 3         |
| 3       | 2         |
| 4       | 1         |
| 5       | 0         |

### When `count` = 10

| Element | CountDown |
|---------|-----------|
| 1       | 4         |
| 2       | 3         |
| 3       | 2         |
| 4       | 1         |
| 5       | 0         |

## Other Notes

It can be used to implement [`SkipLast`](https://morelinq.github.io/2.2/ref/api/html/M_MoreLinq_MoreEnumerable_SkipLast__1.htm):

```c#
var head =
    from x in Enumerable.Range(1, 10)
                        .CountDown(3, (x, cd) => cd is int ? -1 : x)
    where x >= 0
    select x;
```

or [`TakeLast`](https://morelinq.github.io/2.2/ref/api/html/M_MoreLinq_MoreEnumerable_TakeLast__1.htm):

```c#
var tail =
    from x in Enumerable.Range(1, 10)
                        .CountDown(3, (x, cd) => cd is null ? -1 : x)
    where x >= 0
    select x;
```

and more generally anything to do with _x_ tail elements of a sequence (like possibly #456).

**Comment by leandromoh** [comment]
Looks interesting for a new operator!

**Comment by atifaziz** [comment]
Closed via 69b0f1922212ac732f774e18219e716aaed64f8c.
